File: clients/unity/RoomManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace Knet
{
    /// <summary>Result of a successful <see cref="RoomManager.JoinRoomAsync"/> call.</summary>
    public class RoomJoinResult
    {
        public string RoomId;
        public string[] Members;
    }

    /// <summary>
    /// Optional room join/leave/membership layer on top of <see cref="Client"/>,
    /// wrapping the wire protocol defined in <c>roommanager/protocol.go</c>
    /// (Go server side). Mirrors the TS client's <c>RoomManager</c>
    /// (<c>clients/js/src/room-manager.ts</c>).
    ///
    /// <para>
    /// Reconnect caveat: <see cref="Client"/> does not currently persist/resend a
    /// session ID across reconnects, so the server can never treat a reconnect as
    /// a resume (see <c>spec/spec-architecture-room-manager.md</c> CON-005). Every
    /// reconnect is therefore treated as a fresh connection: <see cref="CurrentRooms"/>
    /// is cleared and <see cref="OnRoomsLost"/> fires. The <c>ResumeSync</c> handler
    /// is wired for forward-compatibility but is unreachable with the client as it
    /// exists today.
    /// </para>
    /// </summary>
    public class RoomManager
    {
        private const float DefaultRequestTimeoutSeconds = 10f;

        private readonly Client _client;
        private readonly float _timeoutSeconds;

        private readonly List<string> _rooms = new List<string>();
        private bool _hasConnectedOnce;

        private readonly Dictionary<string, TaskCompletionSource<RoomJoinResult>> _pendingJoins =
            new Dictionary<string, TaskCompletionSource<RoomJoinResult>>();
        private readonly Dictionary<string, TaskCompletionSource<bool>> _pendingLeaves =
            new Dictionary<string, TaskCompletionSource<bool>>();
        private readonly object _pendingLock = new object();

        /// <summary>Raised (main thread) when another client joins a room we're in.</summary>
        public event Action<string, string> OnMemberJoined;

        /// <summary>Raised (main thread) when another client leaves a room we're in.</summary>
        public event Action<string, string> OnMemberLeft;

        /// <summary>Raised (main thread) when another member disconnects (grace period pending).</summary>
        public event Action<string, string> OnMemberDisconnected;

        /// <summary>Raised (main thread) when a previously-disconnected member reconnects.</summary>
        public event Action<string, string> OnMemberReconnected;

        /// <summary>
        /// Raised (main thread) when a reconnect invalidates our room membership
        /// (see class doc / CON-005) — <see cref="CurrentRooms"/> is already empty by then.
        /// </summary>
        public event Action OnRoomsLost;

        /// <summary>Raised (main thread) when another room member sends a message via <see cref="SendMessageAsync"/>.</summary>
        public event Action<string, string, string, string> OnMessage;

        private RoomManager(Client client, float timeoutSeconds)
        {
            _client = client;
            _timeoutSeconds = timeoutSeconds;

            client.On(RoomCommands.JoinAck, payload => HandleJoinAck(Decode<RoomJoinAck>(payload)));
            client.On(RoomCommands.LeaveAck, payload => HandleLeaveAck(Decode<RoomLeaveAck>(payload)));
            client.On(RoomCommands.Error, payload => HandleError(Decode<RoomError>(payload)));
            client.On(RoomCommands.MemberEvent, payload => HandleMemberEvent(Decode<RoomMemberEvent>(payload)));
            client.On(RoomCommands.ResumeSync, payload => HandleResumeSync(Decode<RoomResumeSync>(payload)));
            client.On(RoomCommands.Message, payload => HandleMessage(Decode<RoomMessageEvent>(payload)));

            // currentRooms is intentionally left untouched on OnDisconnected (REQ-019)
            // — resolution happens in HandleConnected on the next reconnect.
            client.OnConnected += HandleConnected;
        }

        /// <summary>Creates a RoomManager bound to <paramref name="client"/>'s room commands.</summary>
        public static RoomManager Attach(Client client, float requestTimeoutSeconds = DefaultRequestTimeoutSeconds)
            => new RoomManager(client, requestTimeoutSeconds);

        /// <summary>Room IDs this client currently believes it's a member of.</summary>
        public IReadOnlyList<string> CurrentRooms => _rooms;

        /// <summary>Requests to join <paramref name="roomId"/>; resolves with the ack's member list.</summary>
        public Task<RoomJoinResult> JoinRoomAsync(string roomId)
        {
            TaskCompletionSource<RoomJoinResult> tcs;
            lock (_pendingLock)
            {
                if (_pendingJoins.ContainsKey(roomId))
                    throw new InvalidOperationException($"join already in flight for room {roomId}");
                tcs = new TaskCompletionSource<RoomJoinResult>(TaskCreationOptions.RunContinuationsAsynchronously);
                _pendingJoins[roomId] = tcs;
            }
            return SendAndAwait(RoomCommands.Join, new RoomJoinRequest { roomId = roomId }, tcs, _pendingJoins, roomId, "join");
        }

        /// <summary>Requests to leave <paramref name="roomId"/>; resolves once acknowledged.</summary>
        public Task LeaveRoomAsync(string roomId)
        {
            TaskCompletionSource<bool> tcs;
            lock (_pendingLock)
            {
                if (_pendingLeaves.ContainsKey(roomId))
                    throw new InvalidOperationException($"leave already in flight for room {roomId}");
                tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
                _pendingLeaves[roomId] = tcs;
            }
            return SendAndAwait(RoomCommands.Leave, new RoomLeaveRequest { roomId = roomId }, tcs, _pendingLeaves, roomId, "leave");
        }

        /// <summary>
        /// Sends an arbitrary message to every other member of <paramref name="roomId"/>
        /// (fire-and-forget; no ack). Requires prior <see cref="JoinRoomAsync"/>.
        /// </summary>
        public Task SendMessageAsync(string roomId, string type, string data = null)
            => _client.SendJsonAsync(RoomCommands.Message, new RoomMessageRequest { roomId = roomId, type = type, data = data });

        private async Task<TResult> SendAndAwait<TResult, TReq>(
            uint commandId, TReq request, TaskCompletionSource<TResult> tcs,
            Dictionary<string, TaskCompletionSource<TResult>> pending, string roomId, string verb)
        {
            try
            {
                await _client.SendJsonAsync(commandId, request).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                lock (_pendingLock) { pending.Remove(roomId); }
                throw new Exception($"failed to send {verb} for room {roomId}: {ex.Message}", ex);
            }

            var timeout = Task.Delay(TimeSpan.FromSeconds(_timeoutSeconds));
            var completed = await Task.WhenAny(tcs.Task, timeout).ConfigureAwait(false);
            if (completed == timeout)
            {
                lock (_pendingLock) { pending.Remove(roomId); }
                throw new TimeoutException($"{verb} room {roomId} timed out after {_timeoutSeconds}s");
            }
            return await tcs.Task.ConfigureAwait(false);
        }

        private void HandleJoinAck(RoomJoinAck ack)
        {
            TaskCompletionSource<RoomJoinResult> tcs;
            lock (_pendingLock) { _pendingJoins.TryGetValue(ack.roomId, out tcs); _pendingJoins.Remove(ack.roomId); }
            tcs?.TrySetResult(new RoomJoinResult { RoomId = ack.roomId, Members = ack.members });
            if (!_rooms.Contains(ack.roomId)) _rooms.Add(ack.roomId);
        }

        private void HandleLeaveAck(RoomLeaveAck ack)
        {
            TaskCompletionSource<bool> tcs;
            lock (_pendingLock) { _pendingLeaves.TryGetValue(ack.roomId, out tcs); _pendingLeaves.Remove(ack.roomId); }
            tcs?.TrySetResult(true);
            _rooms.Remove(ack.roomId);
        }

        private void HandleError(RoomError err)
        {
            var ex = new Exception($"[{err.code}] {err.message}");
            lock (_pendingLock)
            {
                if (_pendingJoins.TryGetValue(err.roomId, out var join))
                {
                    _pendingJoins.Remove(err.roomId);
                    join.TrySetException(ex);
                }
                if (_pendingLeaves.TryGetValue(err.roomId, out var leave))
                {
                    _pendingLeaves.Remove(err.roomId);
                    leave.TrySetException(ex);
                }
            }
        }

        private void HandleMemberEvent(RoomMemberEvent ev)
        {
            switch (ev.type)
            {
                case "joined": OnMemberJoined?.Invoke(ev.roomId, ev.clientId); break;
                case "left": OnMemberLeft?.Invoke(ev.roomId, ev.clientId); break;
                case "disconnected": OnMemberDisconnected?.Invoke(ev.roomId, ev.clientId); break;
                case "reconnected": OnMemberReconnected?.Invoke(ev.roomId, ev.clientId); break;
            }
        }

        private void HandleResumeSync(RoomResumeSync sync)
        {
            // Forward-compatible: unreachable today (see class doc), since the
            // server can only send this after a resumed session, which requires a
            // session id neither client currently persists/resends. If Client
            // gains that, this repopulates rooms after HandleConnected's clear.
            _rooms.Clear();
            if (sync.rooms != null)
                foreach (var r in sync.rooms) _rooms.Add(r.roomId);
        }

        private void HandleMessage(RoomMessageEvent ev)
        {
            OnMessage?.Invoke(ev.roomId, ev.senderId, ev.type, ev.data ?? "");
        }

        private void HandleConnected()
        {
            if (_hasConnectedOnce)
            {
                // A reconnect. No resumable session is possible with the current
                // Client (see class doc / CON-005), so every reconnect is a fresh
                // connection from the server's point of view: drop stale membership.
                _rooms.Clear();
                OnRoomsLost?.Invoke();
            }
            _hasConnectedOnce = true;
        }

        private static T Decode<T>(byte[] payload) => JsonUtility.FromJson<T>(Encoding.UTF8.GetString(payload));
    }
}

File: clients/js/src/room-manager.ts
import type { KNetClient } from "./client.js";
import {
  RoomCommands,
  type RoomError,
  type RoomJoinAck,
  type RoomLeaveAck,
  type RoomMemberEvent,
  type RoomMessageEvent,
  type RoomResumeSync,
} from "./room-protocol.js";

const DEFAULT_REQUEST_TIMEOUT_MS = 10_000;

export interface RoomManagerOptions {
  requestTimeoutMs?: number;
}

export interface RoomJoinResult {
  roomId: string;
  members: string[];
}

interface PendingRequest<T> {
  resolve: (value: T) => void;
  reject: (err: Error) => void;
  timer: ReturnType<typeof setTimeout>;
}

interface RoomManagerEventMap {
  memberJoined: (roomId: string, clientId: string) => void;
  memberLeft: (roomId: string, clientId: string) => void;
  memberDisconnected: (roomId: string, clientId: string) => void;
  memberReconnected: (roomId: string, clientId: string) => void;
  roomsLost: () => void;
  message: (roomId: string, senderId: string, type: string, data: string) => void;
}

const textDecoder = new TextDecoder();

function decodeJson<T>(payload: Uint8Array): T {
  return JSON.parse(textDecoder.decode(payload)) as T;
}

/**
 * Optional room join/leave/membership layer on top of KNetClient, wrapping
 * the wire protocol defined in roommanager/protocol.go (Go server side).
 *
 * Reconnect caveat: KNetClient does not currently persist/resend a session
 * ID across reconnects, so the server can never treat a reconnect as a
 * resume (see spec/spec-architecture-room-manager.md CON-005). Every
 * reconnect is therefore treated as a fresh connection: currentRooms is
 * cleared and `roomsLost` is emitted. The CmdRoomResumeSync handler is
 * wired for forward-compatibility but is unreachable with the client as it
 * exists today.
 */
export class RoomManager {
  private readonly client: KNetClient;
  private readonly timeoutMs: number;

  private rooms = new Set<string>();
  private hasConnectedOnce = false;

  private readonly pendingJoins = new Map<string, PendingRequest<RoomJoinResult>>();
  private readonly pendingLeaves = new Map<string, PendingRequest<void>>();

  private readonly listeners: { [K in keyof RoomManagerEventMap]: Set<RoomManagerEventMap[K]> } = {
    memberJoined: new Set(),
    memberLeft: new Set(),
    memberDisconnected: new Set(),
    memberReconnected: new Set(),
    roomsLost: new Set(),
    message: new Set(),
  };

  constructor(client: KNetClient, opts: RoomManagerOptions = {}) {
    this.client = client;
    this.timeoutMs = opts.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;

    client.onCommand(RoomCommands.JoinAck, (payload) => this.handleJoinAck(decodeJson(payload)));
    client.onCommand(RoomCommands.LeaveAck, (payload) => this.handleLeaveAck(decodeJson(payload)));
    client.onCommand(RoomCommands.Error, (payload) => this.handleError(decodeJson(payload)));
    client.onCommand(RoomCommands.MemberEvent, (payload) => this.handleMemberEvent(decodeJson(payload)));
    client.onCommand(RoomCommands.ResumeSync, (payload) => this.handleResumeSync(decodeJson(payload)));
    client.onCommand(RoomCommands.Message, (payload) => this.handleMessage(decodeJson(payload)));

    // currentRooms is intentionally left untouched on "disconnected" (REQ-019)
    // — resolution happens in handleConnected on the next reconnect.
    client.on("connected", () => this.handleConnected());
  }

  get currentRooms(): readonly string[] {
    return Array.from(this.rooms);
  }

  joinRoom(roomId: string): Promise<RoomJoinResult> {
    if (this.pendingJoins.has(roomId)) {
      return Promise.reject(new Error(`join already in flight for room ${roomId}`));
    }
    const promise = new Promise<RoomJoinResult>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pendingJoins.delete(roomId);
        reject(new Error(`join room ${roomId} timed out after ${this.timeoutMs}ms`));
      }, this.timeoutMs);
      this.pendingJoins.set(roomId, { resolve, reject, timer });
    });
    this.client.sendJson(RoomCommands.Join, { roomId }).catch((err) => {
      const pending = this.pendingJoins.get(roomId);
      if (!pending) return;
      this.pendingJoins.delete(roomId);
      clearTimeout(pending.timer);
      pending.reject(err instanceof Error ? err : new Error(String(err)));
    });
    return promise;
  }

  leaveRoom(roomId: string): Promise<void> {
    if (this.pendingLeaves.has(roomId)) {
      return Promise.reject(new Error(`leave already in flight for room ${roomId}`));
    }
    const promise = new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pendingLeaves.delete(roomId);
        reject(new Error(`leave room ${roomId} timed out after ${this.timeoutMs}ms`));
      }, this.timeoutMs);
      this.pendingLeaves.set(roomId, { resolve, reject, timer });
    });
    this.client.sendJson(RoomCommands.Leave, { roomId }).catch((err) => {
      const pending = this.pendingLeaves.get(roomId);
      if (!pending) return;
      this.pendingLeaves.delete(roomId);
      clearTimeout(pending.timer);
      pending.reject(err instanceof Error ? err : new Error(String(err)));
    });
    return promise;
  }

  sendMessage(roomId: string, type: string, data?: string): Promise<void> {
    return this.client.sendJson(RoomCommands.Message, { roomId, type, data });
  }

  on<K extends keyof RoomManagerEventMap>(event: K, handler: RoomManagerEventMap[K]): void {
    this.listeners[event].add(handler);
  }

  off<K extends keyof RoomManagerEventMap>(event: K, handler: RoomManagerEventMap[K]): void {
    this.listeners[event].delete(handler);
  }

  private emit<K extends keyof RoomManagerEventMap>(event: K, ...args: Parameters<RoomManagerEventMap[K]>): void {
    for (const handler of this.listeners[event]) {
      (handler as (...a: Parameters<RoomManagerEventMap[K]>) => void)(...args);
    }
  }

  private handleJoinAck(ack: RoomJoinAck): void {
    const pending = this.pendingJoins.get(ack.roomId);
    if (pending) {
      this.pendingJoins.delete(ack.roomId);
      clearTimeout(pending.timer);
      pending.resolve({ roomId: ack.roomId, members: ack.members });
    }
    this.rooms.add(ack.roomId);
  }

  private handleLeaveAck(ack: RoomLeaveAck): void {
    const pending = this.pendingLeaves.get(ack.roomId);
    if (pending) {
      this.pendingLeaves.delete(ack.roomId);
      clearTimeout(pending.timer);
      pending.resolve();
    }
    this.rooms.delete(ack.roomId);
  }

  private handleError(err: RoomError): void {
    const join = this.pendingJoins.get(err.roomId);
    if (join) {
      this.pendingJoins.delete(err.roomId);
      clearTimeout(join.timer);
      join.reject(new Error(`[${err.code}] ${err.message}`));
    }
    const leave = this.pendingLeaves.get(err.roomId);
    if (leave) {
      this.pendingLeaves.delete(err.roomId);
      clearTimeout(leave.timer);
      leave.reject(new Error(`[${err.code}] ${err.message}`));
    }
  }

  private handleMemberEvent(ev: RoomMemberEvent): void {
    switch (ev.type) {
      case "joined":
        this.emit("memberJoined", ev.roomId, ev.clientId);
        break;
      case "left":
        this.emit("memberLeft", ev.roomId, ev.clientId);
        break;
      case "disconnected":
        this.emit("memberDisconnected", ev.roomId, ev.clientId);
        break;
      case "reconnected":
        this.emit("memberReconnected", ev.roomId, ev.clientId);
        break;
    }
  }

  private handleMessage(ev: RoomMessageEvent): void {
    this.emit("message", ev.roomId, ev.senderId, ev.type, ev.data ?? "");
  }

  private handleResumeSync(sync: RoomResumeSync): void {
    // Forward-compatible: unreachable today (see class doc), since the
    // server can only send this after a resumed session, which requires a
    // ?session= id neither client currently persists/resends. If a future
    // KNetClient gains that, this repopulates rooms after handleConnected's
    // clear below — correct once ordering allows it.
    this.rooms = new Set(sync.rooms.map((r) => r.roomId));
  }

  private handleConnected(): void {
    if (this.hasConnectedOnce) {
      // A reconnect. No resumable session is possible with the current
      // KNetClient (see class doc / CON-005), so every reconnect is a fresh
      // connection from the server's point of view: drop stale membership.
      this.rooms.clear();
      this.emit("roomsLost");
    }
    this.hasConnectedOnce = true;
  }
}
